### ISSUE TITLE
Self-teaching stack parser

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/error-polyfill#25]
Instead of documenting every environment and trying to write countless parser implementations I thought of writing something general, which can parse every stack string after a short teaching period. Teaching should go by catching a few standard errors and detecting the stack generation patterns by them. These should be native and non-native errors of named, unnamed, etc. functions. Ofc. I'll still need to check many different stack strings, but this would require less effort than maintaining and versioning many different parsers. The other stacktrace project already has many stacktrace examples in its tests, so I could use those by testing the stack parser lib.

**Comment by inf3rno** [comment]
The other project has only saucelabs tests and it does not have tests based on saved stack strings. I am not sure whether I want to test the new parser based on strings or based on actual environment tests. Actually I collected a lot of strings when I implemented the current version, but I ended up testing it with karma instead. On the other hand I cannot test the parser in IE6 and mobile environments, so I need to use strings for those.

**Comment by inf3rno** [comment]
I checked old commits of the other lib. I found some captured stack strings: https://github.com/stacktracejs/stacktrace.js/blob/e8379114b9c849a2c32aba26b9496db607f7e1b9/test/CapturedExceptions.js I'll find the oldest version before using saucelabs. From that I'll have enough example strings to test my own parser.

**Comment by inf3rno** [comment]
I found the last occurence of these captured stacks:
https://github.com/stacktracejs/stacktrace.js/blob/91de59ba95786e3aafa65774c53084cbf1f064ee/test/CapturedExceptions.js

**Comment by inf3rno** [comment]
Let's look for patterns we can use in our parser.

 - there can be multiline frames, we cannot count on single line frames every time, so we need "abc" frames and measure how many line breaks we have. in theory the message cannot contain any line break, but in practice I would not trust that, so I think we'll need at least an "abcd" test to find the borders of a frame and have a frame pattern. the `str.split("\n")` won't work, but we can assume that the actual frame always ends in a line break. yepp I just found a multiline message by old chrome
 - the frame string lines can contain whitespace indentations. by old Opera they contain different size of indentations, so the pattern could looks for that feature too
 - the new line character can depend on the actual OS (`\r\n`, `\n`, etc)
 - the file path can depend on the platform. node+win means `C:\path` instead of `file://C:/path` from a browser or the even better `file://localhost/G:/path` from old Opera. on Linux we have completely different paths, so it is hard to write a pattern for them, but we'll certainly need a "looks like a path" pattern. after that we'll need to normalize the path, which will be hard. I might use the path lib from browserify, but I don't think that will work for any kind of path. we can try to use a relative path, but that does not work by an external script, e.g. loading jquery from their page or another script from a raw github file
 - even the message part can contain some kind of prefix e.g. "Statement on line 42:" by old Opera. and it can contain a postfix as well: "Backtrace:", by old Opera again
 - the first and the last frame strings can be different from the others that's why we need to teach it with an "abcd" stack
 - a frame usually, but not necessarily contains the called function or method, the location, the line and the column. the column is frequently missing, which can be a huge problem if we want to source map a single line uglified, minified js file. yes, everything is in the first line, thanks... sometimes even the function call is not present, or it is just a "()". we might need to normalize even that, but I think that is tricky, because every stack API has a different format for that and the info it contains usually does not overlap. I am not sure I have the tools to normalize it, but I'll give it a try.
 - sometimes the error object contains a sourceURL, but we can't do much with that, because we'd need to know the location of the current js file to have a relative path
 - the line numbers and cols might be different in different browsers, but we might be able to fix that by using evals and adding whitespace padding to the actual code. it will be important by eval to use closures, otherwise it will override the global namespace if the whitespace padding is not working and we use a variable declaration as padding.
 - we can check the properties of the actual error, it can even contain the column data when the stack does not, so that can be very important too

I think we should run the tests first, collect the data and parse all of the stacks and error objects only after that. We should be prepared for throwing strings, etc. in that case there might not be an error object, but that might be environment dependent as well.

**Comment by inf3rno** [comment]
Here is an example:
```
        '    at repl:1:5\n' +
        '    at REPLServer.self.eval (ReadStream.onkeypress:110:21)\n' +
        '    at repl.js:249:20\n' +
        '    at REPLServer.self.eval (Interface.EventEmitter.emit:122:7)\n' +
        '    at Interface.<anonymous> (REPLServer.self.eval:239:12)\n' +
        '    at Interface.EventEmitter.emit (events.js:95:17)\n' +
        '    at readline.js (REPLServer.self:202:10)\n' +
        '    at Interface._line (readline.js:531:8)\n' +
        '    at Interface._ttyWrite (REPLServer.self.eval:760:14)\n' +
        '    at ReadStream.onkeypress (readline.js:99:10)',
```
It can be parsed with 2 regexes:

 - `^.*at (.+)\s+\(([^()]+):(\d+):(\d+)\)[^)]+$` - does not match the `at repl:1:5` line
 - `^.*at ([^()]+):(\d+):(\d+)[^)]+$` - matches the `at repl:1:5` line

What do we learn from it?

**Comment by inf3rno** [comment]
Generating regex-es on the fly is hard. It is even harder to generate regex-es, which match all of the frames, because there can be deviations, e.g. no called function present, no location present, no line/col present, etc... Ofc. we could add a `|` to match all of the frames, but still it is a lot easier to write the postprocessor if it works on a single set of capturing groups.

How would we write a code, which ends up with the 2 regexes in the example?

**Comment by inf3rno** [comment]
I think we should use probe exceptions and compare the generated stacks looking for changes.
 - function name probe: `(function @{name}(){throw new Error;})()`
 - argument name probe: `(function (@{arg}){throw new Error;})(1)`
 - argument value probe: `(function (a,b){throw new Error;})(@{a},@{b})`
 - line probe: `eval("@{verical-padding}throw Error;")`, vertical-padding: fill("\n",i), i=0->9
 - column probe: `eval("@{horizontal-padding}throw Error;")`, horizontal-padding: fill(" ",i), i=0->9
 - file probe: not possible to implement, but maybe there will be enough structural info after the other probes to find paths

The regions don't tend to change in these frame strings, so in this case `func-path-line-col` and if something is no available, then it simply leaves it out: `path-line-col` or `func`. The frame string builder devs usually write lines like `str+=""` instead of  `str=x?(x+y+z):(z+y)`, that's why we can assume that the order does not change. Sometimes they use location value object `location={path,line,col}` and they convert it to string separately: `loc="{path}:{line}:{col}"; str="at {func} ({loc})`. We can look for features like this.

**Comment by inf3rno** [comment]
miről is van szó? hogyan tanulja meg egy ember, hogy mi az a stack trace? valaki megmutatja neki, hogyha hibára fut a kódja, akkor van egy ilyen cucc, ami kiírja, hogy melyik fájlban hol van a hiba, és hogy melyik függvényhívás okozza, és vissza lehet nézni, hogy az adott függvényt mi hívta, és így tovább. ki is lehet próbálni, hogy ha beteszek egy kivétel dobást a kódba, akkor valóban a helyes sort és oszlopot írja e. pl a fentinél látjuk a függvény nevet és az útvonalat, de nem tudjuk, hogy a számoknál a sor vagy az oszlop van e elől. szóval meg kell nézni manuálisan és ellenőrizni, vagy betenni még egy sortörést a fájl elejére, és amelyik változik tőle, az lesz a sor. a gép valamivel ostobább, nem tudja, hogy a függvény neve hogyan néz ki, vagy hogy egy útvonal hogyan néz ki, ezért kell ilyen próbákat használnia. a gond csak annyi, hogy nem tudunk mindenre próbát írni, legalábbis a fájlnévre nem. azt hiszem minden másra igen, ha elég alaposak vagyunk. ha igazán profin csináljuk, akkor sok kicsi atomi próba helyett lehet őket vegyíteni, hogy egy-egy kivétel dobással a lehető legtöbb információt megtudjuk a környezetről. ugyanezt a stratégiát lehet követni egyébként a stack limit esetében is, mert elég csak megnézni, hogy az Error object-en milyen függvények és property-k vannak, és a nevükből meg lehet mondani, hogy melyik az, amelyik a stack limitet tartalmazza, és abból, hogy metódus vagy szám értékű property-e, meg lehet mondani, hogy hogyan kell beállítani. sajna a legtöbb környezetben alapból nincs deklarálva, úgyhogy ha nincs meg, akkor szőnyegbombázás alapon be lehet állítani az ismert property-ket, amik erre szolgálnak, aztán vagy működik, vagy nem. egyébként sem garantált, mert nem minden környezet támogatja. ami érdekes, hogy ezt is lehet próbálni úgy, hogyha alacsonyra állítjuk, pl 1-re vagy 2-re, és megnézzük, hogy változik e a stack hossza.

**Comment by inf3rno** [comment]
data URIs are fine for probing the path: https://stackoverflow.com/questions/49042460/is-there-a-way-to-change-the-path-of-a-function-in-the-stack-trace with base64 encode or with prefix and postfix it is possible to find the location of the path, but this works only in some of the browsers. ie8+ supports data uris, but not by script tags, so that won't work too. maybe there is a workaround for ie, idk. it would be better to have a more general method to probe the path. hmm maybe using iframes?

**Comment by inf3rno** [comment]
I think I'll use node as the main testing environment. I am pretty sure I can mock out the environments I have only stack strings from, old versions of Opera and Firefox. It would be great to test those too. If it will be able to parse multiline frames from old Opera, then I will be sure it is adaptive enough.

**Comment by inf3rno** [comment]
The V8 StackTrace API does not contain function arguments, but in some environments those are part of the function calls. Firefox appear to add primitives. The old Opera appear to copy the actual line to the second frame line. The first frame line contains, whether the error is from the html or from a linked js file. I have to prepare the lib for that. I am not sure how. I might need to install the old Opera to check it. Data URI probably won't work, but maybe iframes do.

Another option to find the words in the frame template and replace each one with `\S+` if they don't match an actual error. For example:

```
"Line 15 of inline#1 script in file://localhost/G:/js/test/functional/testcase1.html\n"
-> Line @{line} of inline#1 script in @{path}\n"
"Line 18 of linked script file://localhost/G:/js/stacktrace.js\n" +
-> does not match
-> .* @{line} of inline#1 script in @{path}\n"
-> still no match
-> .* @{line} .* @{path}\n"
-> match
-> Line @{line} .* @{path}\n"
-> match
-> Line @{line} of .* @{path}\n"
-> match
-> Line @{line} of inline#1 .* @{path}\n"
-> does not match
-> Line @{line} of .* script .* @{path}\n"
-> match
-> Line @{line} of .* script in @{path}\n"
-> does not match
"Line 15 of inline#1 script in file://localhost/G:/js/test/functional/testcase1.html\n"
-> Line @{line} of .* script .* @{path}\n"
-> match -> we can modify the template
```
We need a negative test too, which should not match otherwise using `.*` will work just fine...  Hmm probably we don't need a negative test, we just need to check whether the `line` and `path` are properly parsed. If they are not, then the pattern is too broad. I think this part won't be smart enough to always work. Maybe we can change the `.*` to `\S+` by these words if it matches too, so it won't be too broad. So we need to look for variable regions of the stack string. We won't know what these are, but we can save them as well I think. Hmm maybe I could use the same algorithm to find variable regions as I use for the probes. The only difference here that I don't know what the actual value means by these regions. They are problematic btw if they are close to other variable regions for example here `Line @{line} of .* script .* @{path}` the second `.*` is close to the `path`, which is a variable region and can contain even whitespace in other environments. So there is no way to decide where the unknown variable string ends and where the known variable path starts. I need to find a solution for this problem. The parser must be smart enough to find the beginning of the path based on previous examples. So it should compare previous paths, which is a little bit complicated, because there can be paths like `eval at window.onload (x.html:21), <anonymous>:3:7` or `Unknown script code:8:2`. The function name can contain whitespace as well e.g. `Global code`, or by Opera it is the actual line. So in theory there is no way to properly separate these variable regions from the actual path without understanding what the words mean. Still I think we can find a solution, which will work in the known environments.

**Comment by inf3rno** [comment]
Another topic we haven't discussed is using multiple frame strings. Sometimes there is no function name by chrome and v8: `at @{path}:@{line}:@{col}`. We need to probe these situations as well, otherwise the `at @{func} (@{path}:@{line}:@{col})` won't match. By the probing period it should be possible to add new frame templates and find out whether they are mutually exclusive or whether they both match, so we should use a priority order. After the probing period we can detect the variable regions, but since we cannot probe them, there won't be a way to find out whether we are talking about a new frame template or we should use the same template and add a variable region. I think it will be hard to adapt the existing templates sometimes, but I don't think there is a way to build a new template from scratch without probing. That would require a true AI or even an AI won't be able to solve this without knowing the language and the meaning of a stack string. It is a very interesting problem btw. if I could solve this, then using the same algorithm I would be able to build an AI.

**Comment by inf3rno** [comment]
I think storing the templates as a string is not the way to go. By finding a variable region and mutating the template we'll always check whether the region or the pattern we use can be confused with the special characters we use to describe the known variable regions: func, path, etc. This will be a constant problem, so it will be a lot easier to store template tokens instead of a template string:

```
at @{func} (@{path}:@{line}:@{col})
```
```js
[
    "at ", 
    {name:"func",pattern:".*"}, 
    " (", 
    {name: "path", pattern: ".*"}, 
    ":", 
    {name: "line", pattern: "\d+"}, 
    ":", 
    {name: "col", pattern: "\d+"}, 
    ")"
]
```

Ofc. we'll convert it to regex, but it should be stored this way, because it is a lot easier to mutate it than a string or a regex.

The devtools journal will get a template string I think or maybe a regex with named capturing groups. I know that js does not support that, but who cares, it is for printing...

**Comment by inf3rno** [comment]
We can use typed tokens as well. Another idea by mutation is using a nested array, so if we want to split up a longer variable string, then we won't need to splice the array, which is more demanding and makes it a lot harder to change, split up or restore tokens of the original template. If we found a solution, then we can flatten the arrays.

**Comment by inf3rno** [comment]
I am thinking on a method to parse non-matching frames without mutating the original template. A different strategy to collect more of these frames and compare them with each other. If we are lucky we find the common parts, which we can compare with our already existing template. Probably some of them will be the same and some of them will be different. This way we can do a directed mutation on the template or we can just fork it and create a new template for the frames it did not match.

**Comment by inf3rno** [comment]
If I had an AI with an advanced pattern recognizer, how would it solve the problem?

```js
	r = new TextPatternFinder();
	r.find("type", common->mostly, examples("Error", "SyntaxError", "Exception"));
	r.find("message", common->mostly, examples("Something went wrong.", "Invalid argument."));
	f = r.find("frames[frame]={}", common->mostly, usually single line per item);
	f.find("function", common->mostly, examples(code("js").functionName), special(function=eval -> sub-"frame" is possible)); // chrome nested eval
	f.find("code", rare->always, examples(code("js")), mutex("function")); // old opera
	l = f.find("location{}", common->mostly);
	l.find("path", common->always, examples("x.js", "c:/file.js", "http://a.com/b", "x.html"));
	l.find("line", common->mostly, examples(1, 123, 1032, 10435));
	l.find("column", frequent->mostly, examples(1, 20, 107));
	r.typical("type", "message", "frames");
	f.typical("function", "location").or("location", function);
	l.typical("path","line","column")
	r.possible(unknown params, properties);
	r.possible(not defined params, properties and changing pattern because of that);
	
	p = new Probe(throw error, get stack string from it);
	
	s1 = p(with custom types and messages);
	r.teach(s1[], known(type[], message[]));
	
	s2[] = p(with changing line and column);
	r.teach(s2[], changing(line[], column[]));
	
	s3[] = p(with changing frame count);
	r.teach(s3[], changing(frames.length));
	

	p = new TextParser(r);
	for s of stackstrings
		r.example(s);
		json = p.parse(s, typeconversions);
		log(json);
```

So this is the general approach. I have to design a pattern finder, which can do something similar. It does not have to be this general and intuitive, but it should be able to handle the parameters I am looking for. For example I could give regex patterns instead of examples to recognize the params. I think the important thing that we should define a range where we expect the param to fit in. So for example a line number must be a positive integer and cannot be a text. This way we can identify the params at a certain confidence level.

So we have a lot of expectation about how a stack frame should look like. We should somehow configure the pattern recognizer with these expectations about the information the text contains and about the expected structure of the text, so it will be able to construct effective patterns for the parser to process the text.

we could optimize the parser:
```js
	p = new TextParser(r);
	for s of stackstrings
		if r.unexpected(s)
			r.figureOut(s);
		json = p.parse(s, typeconversions);
		log(json);
```

**Comment by inf3rno** [comment]
An interesting tool I could probably use is rule mining:
http://data-mining.philippe-fournier-viger.com/introduction-to-sequential-rule-mining/

It founds the patterns too and adds a confidence to them. e.g. pattern: `a,b` confidence: `P(b|a) = sup(ab)/sup(a)`. I might be able to use confidence as a filter at some parts of the code. But I am not certain now that I need this tool in the current project. We'll see.

**Comment by inf3rno** [comment]
Meanwhile I read a lot about sequential pattern mining. It is what we need here, but the algorithms aren't that fast for strings. What we need is splitting up the frame to words and special characters and give a numeric id to each of them. So we won't need to run the algorithms for each character.

Check our example:

```js
        '    at repl:1:5\n' +
        '    at REPLServer.self.eval (ReadStream.onkeypress:110:21)\n' +
        '    at repl.js:249:20\n' +
        '    at REPLServer.self.eval (Interface.EventEmitter.emit:122:7)\n' +
        '    at Interface.<anonymous> (REPLServer.self.eval:239:12)\n' +
        '    at Interface.EventEmitter.emit (events.js:95:17)\n' +
        '    at readline.js (REPLServer.self:202:10)\n' +
        '    at Interface._line (readline.js:531:8)\n' +
        '    at Interface._ttyWrite (REPLServer.self.eval:760:14)\n' +
        '    at ReadStream.onkeypress (readline.js:99:10)',
```

What assumptions do we have here?
 - there is some sort of separator between words, usually whitespace, dot, comma, etc. these should have unique ids
 - we don't know if a word is a variable or it is part of the template, so each word should have an id as well
 - numbers are always variables, so they will have the same id
 - unicode characters are not used
 - the line break is the frame string boundary (not always true)

I got this little code:

```js
var stack = 
'    at repl:1:5\n' +
'    at REPLServer.self.eval (ReadStream.onkeypress:110:21)\n' +
'    at repl.js:249:20\n' +
'    at REPLServer.self.eval (Interface.EventEmitter.emit:122:7)\n' +
'    at Interface.<anonymous> (REPLServer.self.eval:239:12)\n' +
'    at Interface.EventEmitter.emit (events.js:95:17)\n' +
'    at readcurrentFrame.js (REPLServer.self:202:10)\n' +
'    at Interface._currentFrame (readcurrentFrame.js:531:8)\n' +
'    at Interface._ttyWrite (REPLServer.self.eval:760:14)\n' +
'    at ReadStream.onkeypress (readcurrentFrame.js:99:10)';

var lineBreakType = 1;
var whiteSpaceType = 2;
var numberType = 3;
var wordType = 4;
var otherType = 5;

var preprocessorPattern = /([\n\r]+)|(\s+)|(\d+)|(\w+)|([^\s\w]+)/gm;

function parser(stackString){
	var stackFrames = [];
	preprocessorPattern.lastIndex = 0;
	var match;
	var currentFrame = [];
	while (match = preprocessorPattern.exec(stackString)){
		var value = match[0];
		for (var matchingGroup = 1; matchingGroup < 5; ++matchingGroup)
			if (match[matchingGroup] !== undefined)
				break;
		var type = matchingGroup;
		if (type == lineBreakType){
			if (currentFrame.length > 0) {
				stackFrames.push(currentFrame);
				currentFrame = [];
			}
			continue;
		}
		var id = getId(type, value);
		currentFrame.push({
			id: id,
			type: type,
			value: value
		});
	}
	return stackFrames;
}

var wordIds = {};
var otherIds = {};
var nextId = 10;

function getId(type, value){
	if (type < 4)
		return type;
	if (type == 4) {
		if (!wordIds.hasOwnProperty(value)){
			wordIds[value] = nextId;
			++nextId;
		}
		return wordIds[value];
	}
	if (!otherIds.hasOwnProperty(value)){
		otherIds[value] = nextId;
		++nextId;
	}
	return otherIds[value];
}

console.log(parser(stack));
```

The result looks like this:

```js
[
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 11, type: 4, value: 'repl' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '1' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '5' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 13, type: 4, value: 'REPLServer' },
    { id: 14, type: 5, value: '.' },
    { id: 15, type: 4, value: 'self' },
    { id: 14, type: 5, value: '.' },
    { id: 16, type: 4, value: 'eval' },
    { id: 2, type: 2, value: ' ' },
    { id: 17, type: 5, value: '(' },
    { id: 18, type: 4, value: 'ReadStream' },
    { id: 14, type: 5, value: '.' },
    { id: 19, type: 4, value: 'onkeypress' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '110' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '21' },
    { id: 20, type: 5, value: ')' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 11, type: 4, value: 'repl' },
    { id: 14, type: 5, value: '.' },
    { id: 21, type: 4, value: 'js' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '249' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '20' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 13, type: 4, value: 'REPLServer' },
    { id: 14, type: 5, value: '.' },
    { id: 15, type: 4, value: 'self' },
    { id: 14, type: 5, value: '.' },
    { id: 16, type: 4, value: 'eval' },
    { id: 2, type: 2, value: ' ' },
    { id: 17, type: 5, value: '(' },
    { id: 22, type: 4, value: 'Interface' },
    { id: 14, type: 5, value: '.' },
    { id: 23, type: 4, value: 'EventEmitter' },
    { id: 14, type: 5, value: '.' },
    { id: 24, type: 4, value: 'emit' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '122' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '7' },
    { id: 20, type: 5, value: ')' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 22, type: 4, value: 'Interface' },
    { id: 25, type: 5, value: '.<' },
    { id: 26, type: 4, value: 'anonymous' },
    { id: 27, type: 5, value: '>' },
    { id: 2, type: 2, value: ' ' },
    { id: 17, type: 5, value: '(' },
    { id: 13, type: 4, value: 'REPLServer' },
    { id: 14, type: 5, value: '.' },
    { id: 15, type: 4, value: 'self' },
    { id: 14, type: 5, value: '.' },
    { id: 16, type: 4, value: 'eval' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '239' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '12' },
    { id: 20, type: 5, value: ')' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 22, type: 4, value: 'Interface' },
    { id: 14, type: 5, value: '.' },
    { id: 23, type: 4, value: 'EventEmitter' },
    { id: 14, type: 5, value: '.' },
    { id: 24, type: 4, value: 'emit' },
    { id: 2, type: 2, value: ' ' },
    { id: 17, type: 5, value: '(' },
    { id: 28, type: 4, value: 'events' },
    { id: 14, type: 5, value: '.' },
    { id: 21, type: 4, value: 'js' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '95' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '17' },
    { id: 20, type: 5, value: ')' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 29, type: 4, value: 'readcurrentFrame' },
    { id: 14, type: 5, value: '.' },
    { id: 21, type: 4, value: 'js' },
    { id: 2, type: 2, value: ' ' },
    { id: 17, type: 5, value: '(' },
    { id: 13, type: 4, value: 'REPLServer' },
    { id: 14, type: 5, value: '.' },
    { id: 15, type: 4, value: 'self' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '202' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '10' },
    { id: 20, type: 5, value: ')' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 22, type: 4, value: 'Interface' },
    { id: 14, type: 5, value: '.' },
    { id: 30, type: 4, value: '_currentFrame' },
    { id: 2, type: 2, value: ' ' },
    { id: 17, type: 5, value: '(' },
    { id: 29, type: 4, value: 'readcurrentFrame' },
    { id: 14, type: 5, value: '.' },
    { id: 21, type: 4, value: 'js' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '531' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '8' },
    { id: 20, type: 5, value: ')' }
  ],
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    { id: 22, type: 4, value: 'Interface' },
    { id: 14, type: 5, value: '.' },
    { id: 31, type: 4, value: '_ttyWrite' },
    { id: 2, type: 2, value: ' ' },
    { id: 17, type: 5, value: '(' },
    { id: 13, type: 4, value: 'REPLServer' },
    { id: 14, type: 5, value: '.' },
    { id: 15, type: 4, value: 'self' },
    { id: 14, type: 5, value: '.' },
    { id: 16, type: 4, value: 'eval' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '760' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '14' },
    { id: 20, type: 5, value: ')' }
  ]
]
```

**Comment by inf3rno** [comment]
Okay, so what is the pattern here?

Let's check it with id only.

```js
[
  [2, 10, 2, 11, 12,  3, 12,  3],
  [2, 10, 2, 13, 14, 15, 14, 16,  2, 17, 18, 14, 19, 12, 3, 12,  3, 20],
  [2, 10, 2, 11, 14, 21, 12, 3, 12,  3],
  [2, 10, 2, 13, 14, 15, 14, 16,  2, 17, 22, 14, 23, 14, 24, 12,  3, 12,  3, 20],
  [2, 10, 2, 22, 25, 26, 27, 2, 17, 13, 14, 15, 14, 16, 12,  3, 12,  3, 20],
  [2, 10, 2, 22, 14, 23, 14, 24,  2, 17, 28, 14, 21, 12, 3, 12,  3, 20],
  [2, 10, 2, 29, 14, 21, 2, 17, 13, 14, 15, 12, 3, 12,  3, 20],
  [2, 10, 2, 22, 14, 30, 2, 17, 29, 14, 21, 12, 3, 12,  3, 20],
  [2, 10, 2, 22, 14, 31,  2, 17, 13, 14, 15, 14, 16, 12, 3, 12,  3, 20]
]

```

**Comment by inf3rno** [comment]
We need to match all frames with 1 pattern, or if that is not possible, then minimize the number of pattern that we use.
The [2, 10, 2, ... , 12, 3, 12, 3, ...] does the job, let's check what it is.

```js
  [
    { id: 2, type: 2, value: '    ' },
    { id: 10, type: 4, value: 'at' },
    { id: 2, type: 2, value: ' ' },
    // ...
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '999' },
    { id: 12, type: 5, value: ':' },
    { id: 3, type: 3, value: '999' },
    // ...
  ]
```
This can be the next layer for parsing after the line breaks. Note that some stacks can contain line breaks. In those cases we have a lot harder problem, because we cannot clearly separate frames. I think that can be solved with pattern mining too, but there are different algorithms for a single string and for multiple strings.

So we can split by line breaks or do it with one pattern like this:
```js
var frameBorderPattern = /(?:^|[\n\r]+)\s+at\s+(.+):(\d+):(\d+)(.*)(?:[\n\r]+|$)/gm;
```
It is not that complicated to generate it based on the upper numbers and it gives something like this in the capturing groups:
```js
['repl', '1', '5', ''],
['REPLServer.self.eval (ReadStream.onkeypress', '110', '21', ')'],
['repl.js', '249', '20', ''],
['REPLServer.self.eval (Interface.EventEmitter.emit', '122', '7', ')'],
['Interface.<anonymous> (REPLServer.self.eval', '239', '12', ')'],
['Interface.EventEmitter.emit (events.js', '95', '17', ')'],
['readcurrentFrame.js (REPLServer.self', '202', '10', ')'],
['Interface._currentFrame (readcurrentFrame.js', '531', '8', ')'],
['Interface._ttyWrite (REPLServer.self.eval', '760', '14', ')'],
['ReadStream.onkeypress (readcurrentFrame.js', '99', '10', ')']
```

**Comment by inf3rno** [comment]
If we check the numbers we have 2 subpatterns here:

```js
  [..., 11, ...],
  [..., 13, 14, 15, 14, 16,  2, 17, 18, 14, 19, ..., 20],
  [..., 11, 14, 21, ...],
  [..., 13, 14, 15, 14, 16,  2, 17, 22, 14, 23, 14, 24, ..., 20],
  [..., 22, 25, 26, 27, 2, 17, 13, 14, 15, 14, 16, ..., 20],
  [..., 22, 14, 23, 14, 24,  2, 17, 28, 14, 21, ..., 20],
  [..., 29, 14, 21, 2, 17, 13, 14, 15, ..., 20],
  [..., 22, 14, 30, 2, 17, 29, 14, 21, ..., 20],
  [..., 22, 14, 31,  2, 17, 13, 14, 15, 14, 16, ..., 20]
```

**Comment by inf3rno** [comment]
How can we find the frame border? I think we need to look for something like this: `(^|1) ... (1|$)` where the `...` cannot be empty or whitespace only and the pattern is the longest with the most matches. We need to be careful with this though, because sometimes the path has a non changing prefix, for example `C:\` or `http://` or `file://`, so the code must understand protocols. This is not a big deal, because we can add a token for protocols if necessary.

**Comment by inf3rno** [comment]
Looks like I already solved the single sequence problem: https://stats.stackexchange.com/a/344708/197098 I'll check the algorithm I described. I'll check maybe I can find other algorithms too.

**Comment by inf3rno** [comment]
It is possible to solve with with sequential pattern mining, probably maximal pattern or such algoritms. The other thing that I can manipulate some of the variables used in the stack string. I need to utilize this, so it is easier to identify these variables in the templates. So I need to override as many I can with random strings and store the random string or number, so I will know what it represents. After that just looking for it in the list of variables of the templates will help a lot. Generating 3-5 stack traces should be enough to mine effectively and cover every possible template. More or less. The recursive ones in Chrome can be a problem, but I don't care about those. Another issue is line break in the message, because some engines don't handle it, but I can deal with these after I have a working model for most browsers. A simple prefixspan mining would be enough. Maybe BIDE. It does not need to be fully optimized if I split up the stack trace by words and special characters.
